### PR TITLE
bugfix: add cache for BannerPlugin to prevent properties like `__SourceMapDevToolData` set in SourceMapDevToolPlugin from losing efficacy when new asset is being created in BannerPlugin

### PR DESCRIPTION
**Pull request by frankLife**
add cache for BannerPlugin to prevent properties like `__SourceMapDevToolData` set in SourceMapDevToolPlugin from losing efficacy when new asset is being created in BannerPlugin

**What kind of change does this PR introduce?**

bugfix

**Did you add tests for your changes?**

yes, I test this logic in my local project

**If relevant, link to documentation update:**

no

**Summary**

I find when I use `BannerPlugin` to add some snippet to asset in watch mode. The building process will get slower obviously.

So I debug into the source code. I find `SourceMapDevToolPlugin` will add a property `__SourceMapDevToolData` to cache the sourcemap result.

But when using BannerPlugin , it would return a new instance based by origin asset.

`compilation.assets[file] = new ConcatSource(banner, "\n", compilation.assets[file]);`

So the asset from `createChunkAssets()` and the new instance both could not have the property `__SourceMapDevToolData` from last compilation process so that the cache will fail.

**Does this PR introduce a breaking change?**

no

**Other information**

I hope this will reviewed or we can talk about how to solve this 🆘

### CODE REVIEW

**Comment by frankLife** [comment]
In AppVeyor tests, I have seen

```
C:\projects\webpack\node_modules\coffee-loader\node_modules\loader-utils\lib\index.js:3
const getOptions = require("./getOptions");
^^^^^
Use of const in strict mode.
 @ (webpack)/test/cases/loaders/coffee-loader/index.js 11:1-40
C:\projects\webpack\node_modules\coffee-loader\node_modules\loader-utils\lib\index.js:3
const getOptions = require("./getOptions");
^^^^^
Use of const in strict mode.
 @ (webpack)/test/cases/loaders/coffee-loader/index.js 2:1-47
C:\projects\webpack\node_modules\coffee-loader\node_modules\loader-utils\lib\index.js:3
const getOptions = require("./getOptions");
^^^^^
Use of const in strict mode.
 @ (webpack)/test/cases/loaders/coffee-loader/index.js 7:1-47
  Error: Errors while compiling:
  
  const getOptions = require("./getOptions")
```

I haven't modify these files. May I should fix these tests ?

**Comment by Kovensky** [comment]
It looks like AppVeyor is testing a very old version of Node, but either its npm is not correctly rejecting a version of loader-utils that doesn't work in that old Node, or loader-utils did not correctly bump the engines in package.json.

**Comment by frankLife** [comment]
So may you @Kovensky  help me change the test config and restart ? I also find `travis-ci` test process is hang on for a long time ..

**Comment by frankLife** [comment]
`travis-ci ` has  taken for more 2 days. We can solve it like restarting it ? @sokra

**Comment by frankLife** [comment]
It looks like the `bundle-loader` 、`coffee-loader`、`url-loader`  using `const` keyword that dosen't match the appveyor test set

```
test log: 

npm WARN engine loader-utils@1.0.2: wanted: {"node":">=4.0.0"} (current: {"node":"0.12.18","npm":"2.15.11"})
npm WARN engine loader-utils@1.0.2: wanted: {"node":">=4.0.0"} (current: {"node":"0.12.18","npm":"2.15.11"})
npm WARN engine loader-utils@1.0.2: wanted: {"node":">=4.0.0"} (current: {"node":"0.12.18","npm":"2.15.11"})
npm WARN engine loader-utils@1.0.2: wanted: {"node":">=4.0.0"} (current: {"node":"0.12.18","npm":"2.15.11"})

```
It seems like the tests run in wrong node environment.

```
test log:

  2) Examples should compile code-splitting-bundle-loader:
     done() invoked with non-Error: C:\projects\webpack\node_modules\bundle-loader\node_modules\loader-utils\lib\index.js:3
const getOptions = require("./getOptions");
^^^^^
Use of const in strict mode.
 @ (webpack)/examples/code-splitting-bundle-loader/example.js 1:0-27
  const getOptions = require("./getOptions");
  ^^^^^
  Use of const in strict mode.
   @ (webpack)/examples/code-splitting-bundle-loader/example.js 1:0-27
      at C:\projects\webpack\test\Examples.test.js:42:13
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:194:14)
      at Compiler.emitRecords (C:\projects\webpack\lib\Compiler.js:282:37)
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:187:11)
      at C:\projects\webpack\lib\Compiler.js:275:11
      at Compiler.afterEmit (C:\projects\webpack\lib\Compiler.js:272:8)
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:267:14)
      at FSReqWrap.oncomplete (fs.js:95:15)
  3) Examples should compile coffee-script:
     done() invoked with non-Error: C:\projects\webpack\node_modules\coffee-loader\node_modules\loader-utils\lib\index.js:3
const getOptions = require("./getOptions");
^^^^^
Use of const in strict mode.
 @ (webpack)/examples/coffee-script/example.js 1:12-29
  const getOptions = require("./getOptions");
  ^^^^^
  Use of const in strict mode.
   @ (webpack)/examples/coffee-script/example.js 1:12-29
      at C:\projects\webpack\test\Examples.test.js:42:13
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:194:14)
      at Compiler.emitRecords (C:\projects\webpack\lib\Compiler.js:282:37)
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:187:11)
      at C:\projects\webpack\lib\Compiler.js:275:11
      at Compiler.afterEmit (C:\projects\webpack\lib\Compiler.js:272:8)
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:267:14)
      at FSReqWrap.oncomplete (fs.js:95:15)
  
  4) Examples should compile component:
     done() invoked with non-Error: C:\projects\webpack\node_modules\url-loader\node_modules\loader-utils\lib\index.js:3
const getOptions = require("./getOptions");
^^^^^
Use of const in strict mode.
 @ (webpack)/~/component-webpack-plugin/~/css-loader!(webpack)/examples/component/components/webpack-a-component/style.css 2:58-80
  const getOptions = require("./getOptions");
  ^^^^^
  Use of const in strict mode.
   @ (webpack)/~/component-webpack-plugin/~/css-loader!(webpack)/examples/component/components/webpack-a-component/style.css 2:58-80
      at C:\projects\webpack\test\Examples.test.js:42:13
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:194:14)
      at Compiler.emitRecords (C:\projects\webpack\lib\Compiler.js:282:37)
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:187:11)
      at C:\projects\webpack\lib\Compiler.js:275:11
      at Compiler.afterEmit (C:\projects\webpack\lib\Compiler.js:272:8)
      at Compiler.<anonymous> (C:\projects\webpack\lib\Compiler.js:267:14)
      at FSReqWrap.oncomplete (fs.js:95:15)
  
  5) TestCases normal compile deduplication-bundle-loader should compile:
     Errors while compiling:

```

and may I fix these `strict mode` mistake ?

@sokra 

Thanks for taking your time to deal with it 👍

**Comment by frankLife** [comment]
May I give a more detail explanation ? This point have a big effect on our daily development with using tool like webpack-dev-server in our team these days, We hope this weakness can be solved as soon as possible 😄

**Comment by TheLarkInn** [comment]
@frankLife are you aware that this PR is running against webpack version 1? Is this already fixed in version 2?

**Comment by frankLife** [comment]
First, Thank you for  response @TheLarkInn 😄 .

Yes, I just test this modification in webpack version 1  when developing our project .
I take a look at the webpack version 2 source code,  and I find there maybe a same question.

In webpack version 2  BannerPlugin.js

```
  chunk.files
    .filter(ModuleFilenameHelpers.matchObject.bind(undefined, options))
    .forEach((file) =>
      compilation.assets[file] = new ConcatSource(
        banner, "\n", compilation.assets[file]
      )
    );
```
`compilation.assets[file] = new ConcatSource(...)` will create a new instance each compilation. 

And in webpack version 2 SourceMapDevToolPlugin.js

```
  if(asset.__SourceMapDevToolData) {
    var data = asset.__SourceMapDevToolData;
    for(var cachedFile in data) {
      this.assets[cachedFile] = data[cachedFile];
      if(cachedFile !== file)
        chunk.files.push(cachedFile);
    }
    return;
  }
```

The `asset.__SourceMapDevToolData` judugement will not match forever if you use Bannerplugin. Because the `__SourceMapDevToolData` property will not exist after the `assets[file]` has been overwrited by a new `ConcatSource` instance when invoking BannerPlugin logic.

**Comment by TheLarkInn** [comment]
In your opinion, would this plugin make more sense as a stand alone entity. Considering if we want to move this to http://github.com/webpack-contrib first. 

Regardless of this question, can you update the pr to compare against v2 webpack instead of v1

**Comment by frankLife** [comment]
It's good to update webpack v2, and I will do this soon.

And if we make this plugin as a stand alone, I think we need to know whether this source comes from the compilation cache object.

like the pr describe ,adding a property `useCache` to `source` object in compilation process [ref](https://github.com/webpack/webpack/pull/4403/files#diff-d6308b00976bdf2edefc0f08dd6e1cafR841).

At last I also want to know whether my understanding about sourcemap cache logic  is right so that I can do it better when deal with it in webpack v2.

File: lib/BannerPlugin.js [hunk 1]
```diff
@@ -16,6 +16,8 @@ function BannerPlugin(banner, options) {
 }
 module.exports = BannerPlugin;
 
+var bannerCache = {};
```
Comment (sokra) [inline lib/BannerPlugin.js]: Use a `WeakMap` for caching.

File: lib/Compilation.js [hunk 2]
```diff
@@ -838,6 +838,7 @@ Compilation.prototype.createChunkAssets = function createChunkAssets() {
 			var usedHash = useChunkHash ? chunkHash : this.fullHash;
 			if(this.cache && this.cache["c" + chunk.id] && this.cache["c" + chunk.id].hash === usedHash) {
 				source = this.cache["c" + chunk.id].source;
+				source.useCache = true;
```
Comment (sokra) [inline lib/Compilation.js]: I don't think you need changes outside the BannerPlugin. Remove this line.

**Comment by sokra** [comment]
Please rebase with PR on master. We won't make any feature changes for webpack 1 anymore.

File: lib/Compilation.js (hunk 2 above)
Comment (frankLife) [inline lib/Compilation.js]: Do your mean whether the asset comes from compilation's cache object is judged by `source.source()` content ? 🤔

Comment (frankLife) [inline lib/Compilation.js]: Adding this property due to performance considering. It doesn't need to invoke other function like `source()`. But is there any better way ?  @sokra

**Comment by frankLife** [comment]
It has removed the `useCache` field in compilation.
The `file` file in BannerPlugin is type of String, but the key of weakmap must be a object so that it is not a right way.

Give a review before transfering the change to webpack v2 @sokra  ?😄

**Comment by frankLife** [comment]
ok, I will submit a pr with update to WeakMap to webpack 2 soon. 
Thank your help

**Pull request closed without merging**